consensus/cryptarchia-sync/core/src/sync.rs

```diff
@@ -0,0 +1,214 @@
+            // Sort the orphan blocks by slot in descending order to minimize backfillings.
+            for (orphan_block, provider_id) in orphan_blocks
+                .iter()
+                .sorted_by_key(|&(_, (slot, _))| std::cmp::Reverse(slot))
+                .map(|(id, (_, provider_id))| (*id, provider_id.clone()))
+            {
+                // Skip the orphan block if it has been processed or rejected
+                // during the previous backfillings.
+                if self.has_block(&orphan_block) || rejected_blocks.contains(&orphan_block) {
+                    continue;
+                }
+
+                match self
+                    .backfill_fork(orphan_block, provider_id, block_fetcher)
+                    .await
+                {
+                    Ok(()) => {}
+                    Err(BackfillError::InvalidBlock {
+                        error,
+                        invalid_suffix,
+                    }) => {
+                        debug!("Invalid blocks found during a backfilling: {error:?}");
+                        rejected_blocks.extend(invalid_suffix);
+                    }
+                    Err(e) => {
+                        return Err(e.into());
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a forward synchronization:
+    /// - Fetch blocks from the peers, starting from the current local tip slot.
+    /// - Process blocks.
+    /// - Gather orphaned blocks whose parent is not in the local block tree.
+    async fn sync_forward(
+        &mut self,
+        start_slot: Slot,
+        block_fetcher: &BlockFetcher,
+        rejected_blocks: &mut HashSet<HeaderId>,
+    ) -> Result<ForwardSyncOutput<BlockFetcher>, Box<dyn Error + Send + Sync>> {
+        let mut max_fetched_slot: Option<Slot> = None;
+        let mut orphan_blocks = HashMap::new();
+
+        info!("Fetching blocks from slot {:?}", start_slot);
+        let mut stream = block_fetcher.fetch_blocks_forward(start_slot).await?;
+        while let Some((block, provider_id)) = stream.next().await {
+            // Update max_fetched_slot if necessary.
+            max_fetched_slot = max_fetched_slot
+                .map(|s| s.max(block.slot()))
+                .or_else(|| Some(block.slot()));
+
+            // Reject blocks that have been rejected in the past
+            // or whose parent has been rejected.
+            let id = block.id();
+            let parent = block.parent();
+            if rejected_blocks.contains(&id) || rejected_blocks.contains(&parent) {
+                rejected_blocks.insert(id);
+                continue;
+            }
+
+            let slot = block.slot();
+            match self.process_block(block).await {
+                Ok(()) => {
+                    debug!("Processed block {id:?}");
+                    orphan_blocks.remove(&id);
+                }
+                Err(CryptarchiaSyncError::ParentNotFound) => {
```

Comment:
How it it possible that when iterating blocks in order you're missing the parent? this should only happen if the starting blocks are not the same, would it be better to handle it separately?